src/protocol/sort/generate_permutation_opt.rs

```diff
@@ -143,7 +143,8 @@ where
 /// # Errors
 pub async fn malicious_generate_permutation_opt<'a, F>(
     sh_ctx: SemiHonestContext<'_, F>,
-    mut sort_keys: impl Iterator<Item = &Vec<Vec<Replicated<F>>>>,
+    // mut sort_keys: impl Iterator<Item = &Vec<Vec<Replicated<F>>>>,
+    mut sort_keys: impl IntoIterator<Item = &Vec<Vec<Replicated<F>>>> + Iterator,
```

Comment:
This is not working. It gives error: the size for values of type `[Vec<replicated::malicious::additive_share::AdditiveShare<F>>]` cannot be known at compilation time
the trait `Sized` is not implemented for `[Vec<replicated::malicious::additive_share::AdditiveShare<F>>]`
all local variables must have a statically known size

I have tried to give "Sized" trait to sort_keys but no use. 
Also, Martin suggested moving impl IntoIterator to "where" clause which is giving issues

Reply:
This maybe one example where Rust compiler gives confusing hints

if your function signature looks like this

```rust
pub async fn malicious_generate_permutation_opt<'a, F>(
    sh_ctx: SemiHonestContext<'_, F>,
    mut sort_keys: impl IntoIterator<Item = &Vec<Vec<Replicated<F>>>>,
) -> Result<(MaliciousValidator<'_, F>, Vec<MaliciousReplicated<F>>), Error> {
   ...
    let first_keys: &Vec<Vec<Replicated<F>>> = sort_keys.next().unwrap();

}
```

then it tells you that it can't do it and suggests to add `+ Iterator` type bound

```bash
 sort_keys: impl IntoIterator<Item = &Vec<Vec<Replicated<F>>>>,
    |                -------------------------------------------------- method `next` not found for this type parameter
...
155 |     let first_keys: &Vec<Vec<Replicated<F>>> = sort_keys.next().unwrap();
    |                                                          ^^^^ method not found in `impl IntoIterator<Item = &Vec<Vec<Replicated<F>>>>`
    |
    = help: items from traits can only be used if the type parameter is bounded by the trait
help: the following traits define an item `next`, perhaps you need to restrict type parameter `impl IntoIterator<Item = &Vec<Vec<Replicated<F>>>>` with one of them:
    |
147 |     sort_keys: impl IntoIterator<Item = &Vec<Vec<Replicated<F>>>> + Iterator,
    |                                                                   ++++++++++
```


the actual problem is that `IntoIterator` is not an iterator but this trait bound implies that this value *can* be converted to an iterator by calling `into_iter`. So the real fix I believe would be 


```rust
pub async fn malicious_generate_permutation_opt<'a, F>(
    sh_ctx: SemiHonestContext<'_, F>,
    sort_keys: impl IntoIterator<Item = &Vec<Vec<Replicated<F>>>>,
) -> Result<(MaliciousValidator<'_, F>, Vec<MaliciousReplicated<F>>), Error>
where
    F: Field,
{
    let mut malicious_validator = MaliciousValidator::new(sh_ctx.clone());
    let mut m_ctx_bit = malicious_validator.context();
    let mut sort_keys = sort_keys.into_iter();
    let first_keys = sort_keys.next().unwrap();
```
```